src/main/scala/org/apache/spark/rdd/OrderedLeftJoinTransformedRDD.scala

```diff
@@ -0,0 +1,46 @@
+package org.apache.spark.rdd
+
+import org.apache.spark._
+import org.broadinstitute.hail.Utils._
+
+import scala.reflect.ClassTag
+
+class OrderedLeftJoinTransformedRDD[T, K1, K2, V1, V2](rdd1: OrderedRDD[T, K1, V1], rdd2: OrderedRDD[T, K2, V2])
+  (implicit ev: (K1) => T, ev2: (K2) => T, tOrd: Ordering[T], kOrd1: Ordering[K1],
+    kOrd2: Ordering[K2], tct: ClassTag[T], kct1: ClassTag[K1],
+    kct2: ClassTag[K2]) extends RDD[(K1, (V1, Option[V2]))](rdd1.sparkContext, Seq(new OneToOneDependency(rdd1),
+  new OrderedDependency(rdd1.orderedPartitioner, rdd2.orderedPartitioner, rdd2)): Seq[Dependency[_]]) {
+
+  override val partitioner: Option[Partitioner] = rdd1.partitioner
+
+  private def p1 = rdd1.orderedPartitioner
+
+  private def p2 = rdd2.orderedPartitioner
+
+  val r2Partitions = rdd2.partitions // compute and store -- getPartitions called on a worker throw a NullPointerExceptione
+
+  def getPartitions: Array[Partition] = rdd1.partitions
+
+  override def getPreferredLocations(split: Partition): Seq[String] = rdd1.preferredLocations(split)
+
+  override def compute(split: Partition, context: TaskContext): Iterator[(K1, (V1, Option[V2]))] = {
+
+    val left = rdd1.iterator(split, context)
+
+    if (left.isEmpty)
+      Iterator.empty
+    else {
+      val first = left.next()
+      val rightStart = p2.getPartition(first._1)
+      val right = (rightStart to OrderedDependency.getDependencies(p1, p2)(split.index)._2)
+        .iterator
+        .flatMap(i => {
+          rdd2.iterator(r2Partitions(i), context)
+        })
+
+      (Iterator(first) ++ left).sortedTransformedLeftJoinDistinct[T, K2, V2](right)
```

Comment:
I would think you want a different interface for the object on the right.  You don't want an iterator, you want something that has an operation "advance until you're at least at this key".  And that should be implemented by binary search on partitions if you're past the current partition.  This way, if left skips forward (e.g., annotating exome with annotations of all SNPs) you don't have to can through everything on the right.